File: src/main/java/me/itzisonn_/meazy/lexer/TokenTypes.java
package me.itzisonn_.meazy.lexer;

import me.itzisonn_.meazy.MeazyMain;
import me.itzisonn_.meazy.registry.Registries;
import org.jspecify.annotations.NullMarked;

/**
 * TokenTypes registrar
 *
 * @see Registries#TOKEN_TYPES
 */
public final class TokenTypes {
    private TokenTypes() {}



    public static TokenType NEW_LINE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("new_line")).getValue();
    }

    public static TokenType END_OF_FILE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("end_of_file")).getValue();
    }



    public static TokenType REQUIRE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("require")).getValue();
    }

    public static TokenType IMPORT() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("import")).getValue();
    }

    public static TokenType VARIABLE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("variable")).getValue();
    }

    public static TokenType FUNCTION() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("function")).getValue();
    }

    public static TokenType CLASS() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("class")).getValue();
    }

    public static TokenType INTERFACE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("interface")).getValue();
    }

    public static TokenType CONSTRUCTOR() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("constructor")).getValue();
    }

    public static TokenType BASE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("base")).getValue();
    }

    public static TokenType IF() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("if")).getValue();
    }

    public static TokenType ELSE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("else")).getValue();
    }

    public static TokenType FOR() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("for")).getValue();
    }

    public static TokenType IN() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("in")).getValue();
    }

    public static TokenType WHILE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("while")).getValue();
    }

    public static TokenType RETURN() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("return")).getValue();
    }

    public static TokenType CONTINUE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("continue")).getValue();
    }

    public static TokenType BREAK() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("break")).getValue();
    }

    public static TokenType IS() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("is")).getValue();
    }

    public static TokenType IS_LIKE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("is_like")).getValue();
    }



    public static TokenType LEFT_PARENTHESIS() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("left_parenthesis")).getValue();
    }

    public static TokenType RIGHT_PARENTHESIS() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("right_parenthesis")).getValue();
    }

    public static TokenType LEFT_BRACE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("left_brace")).getValue();
    }

    public static TokenType RIGHT_BRACE() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("right_brace")).getValue();
    }

    public static TokenType LEFT_BRACKET() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("left_bracket")).getValue();
    }

    public static TokenType RIGHT_BRACKET() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("right_bracket")).getValue();
    }

    public static TokenType COLON() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("colon")).getValue();
    }

    public static TokenType COMMA() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("comma")).getValue();
    }

    public static TokenType DOT() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("dot")).getValue();
    }

    public static TokenType QUESTION() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("question")).getValue();
    }

    public static TokenType QUESTION_DOT() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("question_dot")).getValue();
    }

    public static TokenType QUESTION_COLON() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("question_colon")).getValue();
    }

    public static TokenType ARROW() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("arrow")).getValue();
    }



    public static TokenType ASSIGN() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("assign")).getValue();
    }

    public static TokenType MINUS() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("minus")).getValue();
    }

    public static TokenType POWER() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("power")).getValue();
    }



    public static TokenType AND() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("and")).getValue();
    }

    public static TokenType OR() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("or")).getValue();
    }

    public static TokenType INVERSION() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("inversion")).getValue();
    }



    public static TokenType NULL() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("null")).getValue();
    }

    public static TokenType NUMBER() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("number")).getValue();
    }

    public static TokenType STRING() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("string")).getValue();
    }

    public static TokenType BOOLEAN() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("boolean")).getValue();
    }

    public static TokenType THIS() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("this")).getValue();
    }

    public static TokenType ID() {
        return Registries.TOKEN_TYPES.getEntry(MeazyMain.getDefaultIdentifier("id")).getValue();
    }



    @NativeCanMatch
    @NullMarked
    public static boolean canMatchId(String string) {
        for (TokenType tokenType : TokenTypeSets.KEYWORDS().getTokenTypes()) {
            if (tokenType.getPattern() != null && tokenType.getPattern().matcher(string).matches()) return false;
        }
        return true;
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/lexer/TokenTypes.kt
package me.itzisonn_.meazy.lexer

/**
 * TokenTypes registrar
 */
object TokenTypes {
    private val tokenTypes = mutableSetOf<TokenType>()
    private var hasInitialized = false

    fun add(tokenType: TokenType) {
        require(get(tokenType.id) == null) { "TokenType with id '${tokenType.id}' already exists" }
        tokenTypes += tokenType
    }
    fun get(id: String) = tokenTypes.find { it.id == id }
    fun getAll() = tokenTypes.toList()

    internal fun initialize() {
        check(!hasInitialized) { "TokenTypes have already been initialized" }
        hasInitialized = true

        add(TokenType("new_line", "\n*"))
        add(TokenType("white_space", "(?!\n)\\s", true))
        add(TokenType("end_of_file", null as? Regex))

        add(TokenType("comment", "\\/\\/[^\n]*", true))
        add(TokenType("multi_line_comment", "\\/\\*(?:(?!\\*\\/).)*\\*\\/", true))
        
        add(TokenType("import", "import"))
        add(TokenType("variable", "var|val"))
        add(TokenType("function", "fun"))
        add(TokenType("class", "class"))
        add(TokenType("interface", "interface"))
        add(TokenType("constructor", "constructor"))
        add(TokenType("base", "base"))
        add(TokenType("if", "if"))
        add(TokenType("else", "else"))
        add(TokenType("for", "for"))
        add(TokenType("in", "in"))
        add(TokenType("while", "while"))
        add(TokenType("return", "return"))
        add(TokenType("continue", "continue"))
        add(TokenType("break", "break"))
        add(TokenType("is", "islike|is"))

        add(TokenType("left_parenthesis", "\\("))
        add(TokenType("right_parenthesis", "\\)"))
        add(TokenType("left_brace", "\\{"))
        add(TokenType("right_brace", "\\}"))
        add(TokenType("left_bracket", "\\["))
        add(TokenType("right_bracket", "\\]"))
        add(TokenType("colon", ":"))
        add(TokenType("comma", ","))
        add(TokenType("dot", "\\."))
        add(TokenType("question", "\\?"))
        add(TokenType("question_dot", "\\?\\."))
        add(TokenType("question_colon", "\\?:"))
        add(TokenType("arrow", "->"))

        add(TokenType("assign", "="))
        add(TokenType("plus", "\\+"))
        add(TokenType("minus", "-"))
        add(TokenType("multiply", "\\*"))
        add(TokenType("divide", "\\/"))
        add(TokenType("percent", "%"))
        add(TokenType("power", "\\^"))
        add(TokenType("plus_assign", "\\+="))
        add(TokenType("minus_assign", "-="))
        add(TokenType("multiply_assign", "\\*="))
        add(TokenType("divide_assign", "\\/="))
        add(TokenType("percent_assign", "%="))
        add(TokenType("power_assign", "\\^="))
        add(TokenType("double_plus", "\\+\\+"))
        add(TokenType("double_minus", "--"))

        add(TokenType("and", "&&"))
        add(TokenType("or", "\\|\\|"))
        add(TokenType("inversion", "!"))
        add(TokenType("equals", "=="))
        add(TokenType("not_equals", "!="))
        add(TokenType("greater", ">"))
        add(TokenType("greater_or_equals", ">="))
        add(TokenType("less", "<"))
        add(TokenType("less_or_equals", "<="))
        
        add(TokenType("null", "null"))
        add(TokenType("number", "(0|([1-9][0-9]*))(\\.[0-9]+)?"))
        add(TokenType("string", "\"[^\"\n]*(\")?"))
        add(TokenType("boolean", "true|false"))
        add(TokenType("this", "this"))

        add(
            TokenType(
                "id",
                "[a-zA-Z_][a-zA-Z0-9_]*",
                false
            ) { string ->
                for (tokenType in TokenTypeSets.keywords.getTokenTypes()) {
                    if (tokenType.regex?.matches(string) == true) {
                        return@TokenType false
                    }
                }
                true
            }
        )
    }



    val newLine get() = getNonNull("new_line")
    val endOfFile get() = getNonNull("end_of_file")

    val import get() = getNonNull("import")
    val variable get() = getNonNull("variable")
    val function get() = getNonNull("function")
    val `class` get() = getNonNull("class")
    val `interface` get() = getNonNull("interface")
    val `constructor` get() = getNonNull("constructor")
    val base get() = getNonNull("base")
    val `if` get() = getNonNull("if")
    val `else` get() = getNonNull("else")
    val `for` get() = getNonNull("for")
    val `in` get() = getNonNull("in")
    val `while` get() = getNonNull("while")

    val `return` get() = getNonNull("return")
    val `continue` get() = getNonNull("continue")
    val `break` get() = getNonNull("break")
    val `is` get() = getNonNull("is")

    val leftParenthesis get() = getNonNull("left_parenthesis")
    val rightParenthesis get() = getNonNull("right_parenthesis")
    val leftBrace get() = getNonNull("left_brace")
    val rightBrace get() = getNonNull("right_brace")
    val leftBracket get() = getNonNull("left_bracket")
    val rightBracket get() = getNonNull("right_bracket")
    val colon get() = getNonNull("colon")
    val comma get() = getNonNull("comma")
    val dot get() = getNonNull("dot")
    val question get() = getNonNull("question")
    val questionDot get() = getNonNull("question_dot")
    val questionColon get() = getNonNull("question_colon")
    val arrow get() = getNonNull("arrow")
    
    val assign get() = getNonNull("assign")
    val minus get() = getNonNull("minus")
    val power get() = getNonNull("power")
    
    val and get() = getNonNull("and")
    val or get() = getNonNull("or")
    val inversion get() = getNonNull("inversion")
    
    val `null` get() = getNonNull("null")
    val number get() = getNonNull("number")
    val string get() = getNonNull("string")
    val boolean get() = getNonNull("boolean")
    val `this` get() = getNonNull("this")
    val id get() = getNonNull("id")

    private fun getNonNull(id: String): TokenType {
        return get(id)!!
    }
}
